apps/omg_watcher/lib/omg_watcher/block_validator.ex

```diff
@@ -64,4 +73,93 @@ defmodule OMG.Watcher.BlockValidator do
       end
     end)
   end
+
+  @spec number_of_transactions_within_limit([Transaction.Signed.tx_bytes()]) ::
+          :ok | {:error, atom()}
+  defp number_of_transactions_within_limit(transactions) do
+    case length(transactions) do
+      # A block with a merge transaction only would have no fee transaction.
+      0 ->
+        {:error, :empty_block}
+
+      n when n > @transaction_upper_limit ->
+        {:error, :transactions_exceed_block_limit}
+
+      _ ->
+        :ok
+    end
+  end
+
+  @spec verify_no_duplicate_inputs([Transaction.Recovered.t()]) :: {:ok, map()}
+  defp verify_no_duplicate_inputs(transactions) do
+    Enum.reduce_while(transactions, {:ok, %{}}, fn tx, {:ok, input_set} ->
+      tx
+      |> Map.get(:signed_tx)
+      |> Map.get(:raw_tx)
+      |> Map.get(:inputs, [])
+      # Setting an empty array as default because fee transactions will not have an `input` key.
+      |> scan_for_duplicates({:cont, {:ok, input_set}})
+    end)
+  end
+
+  # Nested reducer executing duplicate verification logic for `verify_no_duplicate_inputs/1`
+  @spec scan_for_duplicates(Transaction.Recovered.t(), {:cont, {:ok, map()}}) :: {:cont, {:ok, map()}}
+  defp scan_for_duplicates(tx_input_set, {:cont, {:ok, input_set}}) do
+    Enum.reduce_while(tx_input_set, {:cont, {:ok, input_set}}, fn input, {:cont, {:ok, input_set}} ->
```

Comment:
[Enum.uniq_by](https://hexdocs.pm/elixir/Enum.html#uniq_by/2) might be faster, not sure.